Extend linux stress to support multivm with multi stress tools

bssrikanth opened the pull request:
Existing linux stress was supporting single vm with single stress
tool. This commit extends that testcase to now support multiple
guests running same/different stress tools inside for given duration

Signed-off-by: Srikanth Aithal <sraithal@linux.vnet.ibm.com>

bssrikanth commented:
Result:
```
JOB ID     : 30ddc43883dead3c0fc1b8a5e7935f2da42f58ba
JOB LOG    : /home/srikanth/avocado-fvt-wrapper/results/job-2018-08-24T08.36-30ddc43/job.log
 (1/1) powerkvm-qemu.linux_stress: PASS (284.04 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 287.70 s
JOB HTML   : /home/srikanth/avocado-fvt-wrapper/results/job-2018-08-24T08.36-30ddc43/results.html
```

balamuruhans commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,17 @@
+    download_type_stressapptest = "git"
+    download_type_stress = "git"
+    stress_cmds_stressapptest = "stressapptest -s 100 -M 256 -m 8 -W"
+    stress_type_vm2 = "stress"
```
this should be,
```
stress_type_vm2 = "stress-ng"
```

bssrikanth replied:
Since params were parsed properly I had left it as-is. Will change to avoid confusion.

balamuruhans commented on generic/tests/linux_stress.py:
```diff
@@ -19,19 +22,41 @@ def run(test, params, env):
+                error = 1
+                logging.error("%s seem to have rebooted during the stress run", vm.name)
+            stress_server = utils_test.VMStress(vm, params)
+            stress_server.unload_stress()
```
more efficient way is to reuse VMstress object created during load_stress_tool() here instead of recreating new object here.

bssrikanth replied:
Done.

balamuruhans commented on generic/tests/linux_stress.py:
```diff
@@ -19,19 +22,41 @@ def run(test, params, env):
+    stress_duration = int(params.get("stress_duration", ""))
+    vms = env.get_all_vms()
+    up_time = {}
+    error = 0
```
let's denote it as `False` and flip it to `True` on failure.

bssrikanth replied:
Taken care.

balamuruhans commented on generic/tests/linux_stress.py:
```diff
@@ -19,19 +22,41 @@ def run(test, params, env):
-    # Please DO define stress_duration to make sure the clean action
-    # being performed, if your case can not be controlled by time,
-    # use utils_test.VMStress() directly
+    stress_duration = int(params.get("stress_duration", ""))
```
default it to 0, else typecast can Trace back with `ValueError: invalid literal for int() with base 10: ''`

bssrikanth replied:
got it. Thanks!

balamuruhans commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,17 @@
+    uninstall_cmds_stress_vm2 = "rm -rf /usr/bin/stress-ng && make clean"
+    download_url_stress_vm2 = "https://github.com/ColinIanKing/stress-ng.git"
+    stress_duration = 10
+    stress_args =""
```
stress-ng_args = "--timer 32 --timer-freq 1000000"

bssrikanth replied:
Yup.

balamuruhans commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,17 @@
+    stress_type_vm2 = "stress"
+    make_cmds_stress_vm2 = "rm -rf /usr/bin/stress-ng && make && cp stress-ng /usr/bin && rm -rf /tmp/stress-*"
+    uninstall_cmds_stress_vm2 = "rm -rf /usr/bin/stress-ng && make clean"
+    download_url_stress_vm2 = "https://github.com/ColinIanKing/stress-ng.git"
```
download_url_stress-ng_vm2 = "https://github.com/ColinIanKing/stress-ng.git"

bssrikanth replied:
Yup.

balamuruhans commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,17 @@
+    download_type_stress = "git"
+    stress_cmds_stressapptest = "stressapptest -s 100 -M 256 -m 8 -W"
+    stress_type_vm2 = "stress"
+    make_cmds_stress_vm2 = "rm -rf /usr/bin/stress-ng && make && cp stress-ng /usr/bin && rm -rf /tmp/stress-*"
```
make_cmds_stress-ng_vm2 = "rm -rf /usr/bin/stress-ng && make && cp stress-ng /usr/bin && rm -rf /tmp/stress-*"

bssrikanth replied:
Yup.

balamuruhans commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,17 @@
+    stress_cmds_stressapptest = "stressapptest -s 100 -M 256 -m 8 -W"
+    stress_type_vm2 = "stress"
+    make_cmds_stress_vm2 = "rm -rf /usr/bin/stress-ng && make && cp stress-ng /usr/bin && rm -rf /tmp/stress-*"
+    uninstall_cmds_stress_vm2 = "rm -rf /usr/bin/stress-ng && make clean"
```
uninstall_cmds_stress-ng_vm2 = "rm -rf /usr/bin/stress-ng && make clean"

bssrikanth replied:
Yup.

bssrikanth commented:
@balamuruhans thanks for the review, I have addressed the review comments. Please have a look at dependent PR https://github.com/avocado-framework/avocado-vt/pull/1726

bssrikanth commented:
@vivianQizhu @sathnaga please review and let me know if you have any comments

bssrikanth commented:
Test result with review changes:
```
JOB ID     : 19eebcfa1613171cdee0289ec23462ae7e1a130a
JOB LOG    : /home/srikanth/avocado-fvt-wrapper/results/job-2018-08-26T23.50-19eebcf/job.log
 (1/1) powerkvm-qemu.linux_stress: PASS (286.02 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 289.84 s
JOB HTML   : /home/srikanth/avocado-fvt-wrapper/results/job-2018-08-26T23.50-19eebcf/results.html
```

```2018-08-26 23:51:06,259 remote           L0480 DEBUG| SCP process terminated with status 0
2018-08-26 23:51:06,267 remote           L0871 INFO | Copy file from /var/tmp/avocado_FebNWx/stressapptest.git to 192.168.122.169:/home, elapsed time: 0.896224021912
2018-08-26 23:51:06,268 __init__         L2040 INFO | make and install stressapptest
2018-08-26 23:51:06,268 client           L1126 DEBUG| Sending command: cd /home/stressapptest.git/;./configure && make && make install
2018-08-26 23:51:28,198 client           L1126 DEBUG| Sending command: echo $?
2018-08-26 23:51:28,299 client           L1162 DEBUG| Sending command (safe): cd /home/stressapptest.git/
2018-08-26 23:51:28,400 __init__         L1931 INFO | Launch stress with command: nohup stressapptest -s 100 -M 256 -m 8 -W  > /dev/null &
2018-08-26 23:51:28,400 client           L1162 DEBUG| Sending command (safe): nohup stressapptest -s 100 -M 256 -m 8 -W  > /dev/null &
2018-08-26 23:51:31,005 utils_misc       L0683 DEBUG| wait for stress app to start (2.002014 secs)
2018-08-26 23:51:31,005 client           L1126 DEBUG| Sending command: pidof -s stressapptest
2018-08-26 23:51:31,137 client           L1126 DEBUG| Sending command: echo $?
--
2018-08-26 23:51:36,985 remote           L0480 DEBUG| SCP process terminated with status 0
2018-08-26 23:51:36,986 remote           L0871 INFO | Copy file from /var/tmp/avocado_FebNWx/stressapptest.git to 192.168.122.106:/home, elapsed time: 0.905330896378
2018-08-26 23:51:36,987 __init__         L2040 INFO | make and install stressapptest
2018-08-26 23:51:36,987 client           L1126 DEBUG| Sending command: cd /home/stressapptest.git/;./configure && make && make install
2018-08-26 23:52:00,197 client           L1126 DEBUG| Sending command: echo $?
2018-08-26 23:52:00,298 client           L1162 DEBUG| Sending command (safe): cd /home/stressapptest.git/
2018-08-26 23:52:00,399 __init__         L1931 INFO | Launch stress with command: nohup stressapptest -s 100 -M 256 -m 8 -W  > /dev/null &
2018-08-26 23:52:00,399 client           L1162 DEBUG| Sending command (safe): nohup stressapptest -s 100 -M 256 -m 8 -W  > /dev/null &
2018-08-26 23:52:03,006 utils_misc       L0683 DEBUG| wait for stress app to start (2.002061 secs)
2018-08-26 23:52:03,008 client           L1126 DEBUG| Sending command: pidof -s stressapptest
2018-08-26 23:52:03,200 client           L1126 DEBUG| Sending command: echo $?
--
2018-08-26 23:52:13,612 remote           L0480 DEBUG| SCP process terminated with status 0
2018-08-26 23:52:13,613 remote           L0871 INFO | Copy file from /var/tmp/avocado_FebNWx/stress-ng.git to 192.168.122.11:/home, elapsed time: 1.72914099693
2018-08-26 23:52:13,614 __init__         L2040 INFO | make and install stress-ng
2018-08-26 23:52:13,614 client           L1126 DEBUG| Sending command: cd /home/stress-ng.git/;rm -rf /usr/bin/stress-ng && make && cp stress-ng /usr/bin && rm -rf /tmp/stress-*
2018-08-26 23:52:58,121 client           L1126 DEBUG| Sending command: echo $?
2018-08-26 23:52:58,223 client           L1162 DEBUG| Sending command (safe): cd /home/stress-ng.git/
2018-08-26 23:52:58,325 __init__         L1931 INFO | Launch stress with command: nohup stress-ng --timer 32 --timer-freq 1000000  > /dev/null &
2018-08-26 23:52:58,326 client           L1162 DEBUG| Sending command (safe): nohup stress-ng --timer 32 --timer-freq 1000000  > /dev/null &
2018-08-26 23:53:00,969 utils_misc       L0683 DEBUG| wait for stress app to start (2.002102 secs)
2018-08-26 23:53:00,969 client           L1126 DEBUG| Sending command: pidof -s stress-ng
2018-08-26 23:53:01,267 client           L1126 DEBUG| Sending command: echo $?
```

sathnaga commented on generic/tests/linux_stress.py:
```diff
@@ -19,19 +22,42 @@ def run(test, params, env):
+
+    for vm in vms:
+        try:
+            uptime = vm.uptime()
```
you can make this uptime check in try block as you suspect of below network issue, can be checked in the same block, as it will any way fail out here itself incase of network not reachable.

bssrikanth replied:
Yes you are correct. Thanks!

bssrikanth commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,17 @@
+    uninstall_cmds_stress-ng_vm2 = "rm -rf /usr/bin/stress-ng && make clean"
+    download_url_stress-ng_vm2 = "https://github.com/ColinIanKing/stress-ng.git"
+    stress_duration = 10
+    stress_args =""
```
Yup.. thanks for pointing it!

bssrikanth commented:
@balamuruhans @sathnaga thanks for the review. I have incorporated all comments. @vivianQizhu if you have any comments please let me know

balamuruhans commented:
@sathnaga @vivianQizhu, request for your review and acknowledgement.

vivianQizhu commented on generic/tests/cfg/linux_stress.cfg:
```diff
@@ -1,6 +1,19 @@
     type = linux_stress
-    stress_duration = 60
-    # Please specify below params for stress test, or use default value as shown
-    # download_url = http://people.seas.harvard.edu/~apw/stress/stress-1.0.4.tar.gz
```
Can we keep the original test app `stress-1.0.4.tar.gz` as `- @default:` and below `stressapptest` as another variants `- sressapptest:`? Cause we might need the original one for some of our test cases.

bssrikanth replied:
:+1:  done.

bssrikanth replied:
done

balamuruhans commented:
Thanks all, let's get it in.

vivianQizhu commented:
@bssrikanth @balamuruhans 
Hi, sorry for my careless when ack it, but this brings a regression on qemu vm test. We have no uptime() in qemu_vm.VM() object. That's only supported by libvirt_vm.
I would prefer revert this commit since we don't have multi vm test requirement in qemu test, and you might create such similar test case in tp-libvirt. 
If you do need test this tp-qemu case by libvirt vm, we can either use utils_misc.get_uptime() instead or remove that condition check(start load stress directly).
What do you say? Thanks.

balamuruhans commented:
@vivianQizhu, uptime() can be used to validate VM reboot during a test. I have implemented uptime() for qemu_vm in https://github.com/avocado-framework/avocado-vt/pull/1757 help reviewing. 

Thanks for bringing it up.

bssrikanth commented:
@vivianQizhu As @luckyh has suggested we could use utils_misc.get_uptime(session) directly in linux_stress test.

bssrikanth commented:
@vivianQizhu .. @balamuruhans might send that fix.. if you still want to revert this commit and have same in libvirt let me know.

vivianQizhu commented:
@bssrikanth @balamuruhans I am okay for using utils_misc.get_uptime() instead. Thanks.
